clang/lib/StaticAnalyzer/Checkers/StdLibraryFunctionsChecker.cpp

```diff
@@ -436,9 +436,9 @@ class StdLibraryFunctionsChecker
                                llvm::raw_ostream &Out) const override;
 
     ValueConstraintPtr negate() const override {
-      NotNullBufferConstraint Tmp(*this);
+      BufferNullnessConstraint Tmp(*this);
       Tmp.CannotBeNull = !this->CannotBeNull;
```

Comment:
The negation of the constraint does not fully match my expectation: this constraint is saying `size == 0 || buf != 0`. So the negation is suppose to be `size != 0 && buf == 0`. The implementation looks more like `size == 0 || buf == 0` to me.

In general, how is the negation of a constraint used? I imagine that sometimes the engine needs to split the path by assuming '_C_' and '!_C_' for a constraint '_C_'? If the negation can be used as an assumption for a path, we should be serious about its correctness. Additionally, `describe` and `describeArgumentValue` need to report differently depending on whether the constraint is negated or not.

Reply:
The negation of a constraint `C` is only used to determine if the constraint holds at a state.  It checks that applying `C` results in a feasible state and applying `!C` results in a infeasible one. 

The current negation implementation of `NotNullBufferConstraint` is kind of "weaker" than what it suppose to be. (Formally, the "weaker than" relation doesn't exist between them.)  In other words, applying `!C` may not result in an infeasible state when it is suppose to happen.  Consequently,  the checker can miss the chance of determining a constraint holds at a state.  But I suspect this is some rare corner case.

My conclusion is to make no change but some comments about this. @balazske

Reply:
Oh, I somehow thought the checker passes a pre-condition `C` only when `State->assume(C) && !State->assume(!C))` but the reality is it reports a violation only when `State->assume(!C) && !State->assume(C)`.

Yes, you are right. This is fine.